# Business Problem Understanding

## Project Overview

Bank XYZ is a financial institution with a large customer base consisting primarily of liability customers (depositors). While the bank has successfully attracted customers who maintain savings and deposit accounts, it aims to increase its asset customer base by encouraging more customers to take loans and other credit products.

To achieve this objective, the bank recently conducted a marketing campaign targeting existing customers. However, the campaign resulted in a relatively low conversion rate, leading to inefficient utilization of marketing resources and budget.

As part of the bank's digital transformation strategy, management seeks a data-driven solution that can identify customers who are most likely to accept loan offers. By targeting only high-potential customers, the bank can improve campaign effectiveness, increase conversion rates, and maximize return on marketing investment.

## Business Objective

The primary objective of this project is to develop a machine learning classification model capable of identifying customers who are likely to convert from liability customers to asset customers by accepting a loan offer.

The model will help the marketing team:

Improve campaign targeting efficiency.
Increase loan conversion rates.
Optimize marketing expenditures.
Support data-driven decision-making.
Enhance customer acquisition strategies.

## Business Impact

A successful predictive model can provide several benefits to Bank XYZ:

#### Improved Marketing Efficiency
- Rather than targeting the entire customer base, the bank can focus marketing efforts on customers with a higher probability of conversion.

#### Increased Revenue
- Loan products generate interest income, which contributes significantly to the bank's profitability.

#### Reduced Customer Acquisition Cost
- Targeted campaigns help reduce unnecessary marketing expenses by minimizing outreach to unlikely converters.

#### Better Customer Experience
- Customers receive offers that are more relevant to their financial profile and needs.

## Machine Learning Problem Statement

This project can be formulated as a binary classification problem where the goal is to predict whether a customer is likely to have a loan on their credit card.

#### Target Variable

| Variable   | Description                                                   |
|------------|---------------------------------------------------------------|
| LoanOnCard | Indicates whether the customer has a loan on their credit card |

#### Target Classes

| Value | Meaning                          |
|--------|----------------------------------|
| 0      | Customer does not have a loan    |
| 1      | Customer has a loan              |

The model will learn patterns from customer demographics, spending behavior, banking relationships, and financial attributes to predict the likelihood of loan conversion.

## Expected Project Outcome

The final output of this project will be a trained classification model that can predict whether a customer is likely to become a borrower.

Several classification algorithms will be explored, including:
- Logistic Regression
- Weighted Logistic Regression
- Naive Bayes
- Support Vector Machine
- Decision Tree Classifier
- Random Forest Classifier

The best-performing model will be selected based on business-relevant metrics such as recall, precision, F1-score, confusion matrix, ROC-AUC, and accuracy.

Because the target variable is imbalanced, class imbalance handling techniques such as imblearn, SMOTE, and hybrid resampling will also be considered.

## Business Success Criteria

The project will be considered successful if the final model can correctly identify a high proportion of potential borrowers.

For this business problem, recall is especially important because failing to identify a potential borrower may result in missed revenue opportunities.

However, precision is also important because contacting too many unlikely customers may waste marketing budget.

Therefore, the final model should balance:
- High recall
- Good precision
- Strong F1-score
- Strong ROC-AUC
- Business interpretability

# Data Collection

## Data Source

The dataset was provided as part of a banking analytics case study and contains customer-level information from Bank XYZ.

The data is distributed across two separate files containing customer demographic, financial, and banking relationship attributes.

## Dataset Overview

The dataset is split into two CSV files using the common identifier **ID**.

| Dataset | Rows | Columns | Description |
|----------|------|---------|-------------|
| Data1 | 5,000 | 8 | Customer demographic and spending-related information |
| Data2 | 5,000 | 7 | Banking product ownership and target variable information |

Both datasets contain the common column ID, which will be used to merge the files into one final analytical dataset.

## Data Dictionary

| Feature | Description |
|----------|-------------|
| ID | Unique customer identifier |
| Age | Customer's approximate age |
| CustomerSince | Customer relationship duration with the bank; unit is masked |
| HighestSpend | Customer's highest spend in a single transaction; unit is masked |
| ZipCode | Customer's zip code |
| HiddenScore | Masked score associated with the customer |
| MonthlyAverageSpend | Customer's monthly average spend; unit is masked |
| Level | Masked customer level assigned by the bank |
| Mortgage | Customer's mortgage value; unit is masked |
| Security | Indicates whether the customer holds security assets with the bank |
| FixedDepositAccount | Indicates whether the customer has a fixed deposit account |
| InternetBanking | Indicates whether the customer uses internet banking |
| CreditCard | Indicates whether the customer uses the bank's credit card |
| LoanOnCard | Target variable indicating whether the customer has a loan on card |

### Feature Categories

| Category | Features |
|-----------|----------|
| Customer Demographics | Age, ZipCode |
| Customer Relationship | CustomerSince, Level |
| Spending Behavior | HighestSpend, MonthlyAverageSpend |
| Financial Attributes | Mortgage, HiddenScore |
| Banking Products & Services | Security, FixedDepositAccount, InternetBanking, CreditCard |
| Identifier | ID |
| Target Variable | LoanOnCard |

### Target Variable

| Variable | Description |
|-----------|-------------|
| LoanOnCard | Indicates whether the customer has a loan on their credit card |

### Target Classes

| Value | Meaning |
|--------|---------|
| 0 | Customer does not have a loan |
| 1 | Customer has a loan |

# Data Loading and Integration

In [1]:
# Import Required Libraries

# Path
from pathlib import Path

# Data Manipulation
import numpy as np
import pandas as pd

# Warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
from pathlib import Path
import pandas as pd


def load_dataset(file_path):
    return pd.read_csv(file_path)


DATA_DIR = Path("../data/raw")

data1 = load_dataset(DATA_DIR / "Data1.csv")
data2 = load_dataset(DATA_DIR / "Data2.csv")

In [3]:
# Inspect Dataset Dimensions

print(f"Data1 Shape: {data1.shape}")
print(f"Data2 Shape: {data2.shape}")

Data1 Shape: (5000, 8)
Data2 Shape: (5000, 7)


In [4]:
# Preview Raw Datasets

display(data1.head())
display(data2.head())

,ID,Age,CustomerSince,HighestSpend,ZipCode,HiddenScore,MonthlyAverageSpend,Level
0,1,25,1,49,91107,4,1.6,1
1,2,45,19,34,90089,3,1.5,1
2,3,39,15,11,94720,1,1.0,1
3,4,35,9,100,94112,1,2.7,2
4,5,35,8,45,91330,4,1.0,2


,ID,Mortgage,Security,FixedDepositAccount,InternetBanking,CreditCard,LoanOnCard
0,1,0,1,0,0,0,NaN
1,2,0,1,0,0,0,NaN
2,3,0,0,0,0,0,NaN
3,4,0,0,0,0,0,NaN
4,5,0,0,0,0,1,NaN


In [5]:
# Check for Duplicate IDs

print("Data1 Duplicate IDs:", data1['ID'].duplicated().sum())
print("Data2 Duplicate IDs:", data2['ID'].duplicated().sum())

Data1 Duplicate IDs: 0
Data2 Duplicate IDs: 0


In [6]:
# Check Common Customers

common_ids = set(data1['ID']).intersection(set(data2['ID']))

print(f"Customers in Data1: {data1['ID'].nunique()}")
print(f"Customers in Data2: {data2['ID'].nunique()}")
print(f"Common Customers: {len(common_ids)}")

Customers in Data1: 5000
Customers in Data2: 5000
Common Customers: 5000


In [7]:
# Merge Datasets Using Customer ID

# Inner join is used because both files should contain matching customer IDs.

cust_data = pd.merge(
    data1,
    data2,
    on="ID",
    how="inner",
    indicator=True
)

cust_data['_merge'].value_counts()

_merge
both          5000
left_only        0
right_only       0
Name: count, dtype: int64

In [8]:
# Drop temporary merge status column generated during dataset integration
cust_data.drop(columns='_merge', inplace=True)

In [9]:
# Validate Final Merged Dataset

print(f"Final Dataset Shape: {cust_data.shape}")

display(cust_data.head())

Final Dataset Shape: (5000, 14)


,ID,Age,CustomerSince,HighestSpend,ZipCode,HiddenScore,MonthlyAverageSpend,Level,Mortgage,Security,FixedDepositAccount,InternetBanking,CreditCard,LoanOnCard
0,1,25,1,49,91107,4,1.6,1,0,1,0,0,0,NaN
1,2,45,19,34,90089,3,1.5,1,0,1,0,0,0,NaN
2,3,39,15,11,94720,1,1.0,1,0,0,0,0,0,NaN
3,4,35,9,100,94112,1,2.7,2,0,0,0,0,0,NaN
4,5,35,8,45,91330,4,1.0,2,0,0,0,0,1,NaN


In [10]:
# Check Whether All Customer IDs Were Successfully Matched

print(f"Unique IDs in Data1: {data1['ID'].nunique()}")
print(f"Unique IDs in Data2: {data2['ID'].nunique()}")
print(f"Unique IDs in Final Dataset: {cust_data['ID'].nunique()}")

Unique IDs in Data1: 5000
Unique IDs in Data2: 5000
Unique IDs in Final Dataset: 5000


# Saving Merged Dataset

In [11]:
# Save Merged Customer Dataset

# Create directory if it does not exist
interim_path = Path("../data/processed")
interim_path.mkdir(parents=True, exist_ok=True)

# Output file path
output_file = interim_path / "01_merged_customer_data.csv"

# Save dataset
cust_data.to_csv(
    output_file,
    index=False
)

print(f"Merged dataset saved successfully at: {output_file}")

Merged dataset saved successfully at: ..\data\interim\01_merged_customer_data.csv


# Data Understanding

In [12]:
# Feature Data Types Summary

dtype_summary = pd.DataFrame({
    "Feature": cust_data.columns,
    "Data Type": cust_data.dtypes.values,
    "Non-Null Count": cust_data.count().values,
    "Missing Values": cust_data.isna().sum().values
})

dtype_summary

,Feature,Data Type,Non-Null Count,Missing Values
0,ID,int64,5000,0
1,Age,int64,5000,0
2,CustomerSince,int64,5000,0
3,HighestSpend,int64,5000,0
4,ZipCode,int64,5000,0
5,HiddenScore,int64,5000,0
6,MonthlyAverageSpend,float64,5000,0
7,Level,int64,5000,0
8,Mortgage,int64,5000,0
9,Security,int64,5000,0


## Observation
- The dataset contains 5,000 customer records and 14 features.
- No missing values are present in the predictor variables.
- The target variable contains 20 missing observations (0.40%).
- All features are numerical, simplifying preprocessing.
- ID will be excluded from model training.
- Potential challenges include class imbalance, outliers, and high-cardinality zip codes.
- The dataset is generally clean and suitable for predictive modeling.

## Missing Value Assessment

In [13]:
# Missing Value Analysis


missing_summary = pd.DataFrame({
    "Missing Values": cust_data.isnull().sum(),
    "Missing Percentage": (
        cust_data.isnull().mean() * 100
    ).round(2)
})

missing_summary[missing_summary["Missing Values"] > 0]

,Missing Values,Missing Percentage
LoanOnCard,20,0.4


## Observation
- LoanOnCard contains 20 missing values (0.40%).
- No predictor variables contain missing values.

Missing target observations will be investigated and handled during the Data Cleaning & Preprocessing phase.

# Duplicate Record Analysis

In [14]:
# Duplicate Record Analysis

cust_data.duplicated().sum()

0

## Observation
- No duplicate records were identified.
- Each observation represents a unique customer record.

## Unique Value Analysis

In [15]:
# Unique Value Analysis

unique_values = pd.DataFrame({
    "Feature": cust_data.columns,
    "Unique Values": cust_data.nunique().values
}).sort_values(
    by="Unique Values",
    ascending=False
)

unique_values

,Feature,Unique Values
0,ID,5000
4,ZipCode,467
8,Mortgage,347
3,HighestSpend,162
6,MonthlyAverageSpend,108
2,CustomerSince,47
1,Age,45
5,HiddenScore,4
7,Level,3
9,Security,2


## Observation
- Every customer record is unique.
- No identifier leakage except the ID column.
- Financial variables appear rich in behavioral information.
- HiddenScore and Level are promising predictive features.
- Binary variables are already suitable for machine learning.
- ZipCode requires further investigation due to high cardinality.

# Feature Categorization

In [16]:
# Target Variable
target_col = "LoanOnCard"

# Identifier Variable
id_cols = ["ID"]

# Geographic Variable
location_cols = ["ZipCode"]

# Numerical Features
num_cols = [
    "Age",
    "CustomerSince",
    "HighestSpend",
    "MonthlyAverageSpend",
    "Mortgage"
]

# Ordinal Features
ordinal_cols = [
    "Level",
    "HiddenScore"
]

# Binary Features
binary_cols = [
    "Security",
    "FixedDepositAccount",
    "InternetBanking",
    "CreditCard"
]

In [17]:
# Validate Feature Classification

classified_cols = (
    id_cols
    + location_cols
    + num_cols
    + ordinal_cols
    + binary_cols
    + [target_col]
)

uncategorized_cols = [
    col for col in cust_data.columns
    if col not in classified_cols
]

print("Uncategorized Columns:", uncategorized_cols)

Uncategorized Columns: []


## Feature Categorization Summary

The variables were grouped according to their business meaning and analytical characteristics to facilitate subsequent stages of data exploration, preprocessing, feature engineering, and model development.

### Identifier Variable

The ID feature uniquely identifies each customer and does not provide predictive information. Therefore, it will be excluded from model training.

### Geographic Variable

The ZipCode feature represents customer location information. Due to its high cardinality (467 unique values), its predictive usefulness will be evaluated during exploratory data analysis before deciding whether to retain, transform, or remove it.

### Numerical Variables

The numerical features capture customer demographics and spending behavior:
- Age
- CustomerSince
- HighestSpend
- MonthlyAverageSpend
- Mortgage

These variables will be analyzed for:
- Distribution patterns
- Outliers
- Skewness
- Correlation
- Predictive importance

### Ordinal Variables

The variables Level and HiddenScore appear to represent customer rankings assigned by the bank.

Because these features possess a natural ordering, they will be treated as ordinal variables during analysis.

### Binary Variables

The following features represent yes/no banking product ownership indicators:
- Security
- FixedDepositAccount
- InternetBanking
- CreditCard

These variables are already encoded numerically and are directly suitable for machine learning algorithms.

### Target Variable

The target variable is:

LoanOnCard

where:

| Value | Class Description |
|--------|------------------|
| 0 | Customer does not have a loan on their credit card |
| 1 | Customer has a loan on their credit card |

This project is therefore formulated as a binary classification problem.

# Data Quality Assessment: Range and Validity Checks

## Validate Numerical Feature Ranges

In [18]:
# Range and Validity Checks

# Verify that numerical features fall within reasonable business and real-world limits.

range_checks = {
    
    # Customer age should be at least 18 years
    "Age < 18": (cust_data["Age"] < 18).sum(),

    # Ages above 100 are considered unrealistic
    "Age > 100": (cust_data["Age"] > 100).sum(),

    # Customer tenure cannot be negative
    "CustomerSince < 0": (
        cust_data["CustomerSince"] < 0
    ).sum(),

    # Highest spending amount cannot be negative
    "HighestSpend < 0": (
        cust_data["HighestSpend"] < 0
    ).sum(),

    # Monthly average spending cannot be negative
    "MonthlyAverageSpend < 0": (
        cust_data["MonthlyAverageSpend"] < 0
    ).sum(),

    # Mortgage amount cannot be negative
    "Mortgage < 0": (
        cust_data["Mortgage"] < 0
    ).sum()
}

range_check_summary = pd.DataFrame({
    "Validation Rule": range_checks.keys(),
    "Invalid Records": range_checks.values()
})

range_check_summary

,Validation Rule,Invalid Records
0,Age < 18,0
1,Age > 100,0
2,CustomerSince < 0,52
3,HighestSpend < 0,0
4,MonthlyAverageSpend < 0,0
5,Mortgage < 0,0


## Observations

The range validation exercise revealed that:
- No unrealistic age values were detected.
- No negative spending values were observed.
- No negative mortgage values were identified.
- However, 52 observations contain negative values in the CustomerSince feature.

## CustomerSince Validation

A total of 52 observations were found to contain negative values in the CustomerSince variable.

Since the dataset documentation indicates that the feature has been masked by the bank, it is not possible to determine whether these values represent data entry errors or valid encoded information.

Therefore, no corrective action is taken during the Data Understanding phase. The variable will be investigated further during exploratory data analysis to determine whether the negative values require treatment before model development.

## Ordinal Feature Validation

In [19]:
# Ordinal Feature Summary

ordinal_summary = pd.DataFrame({
    "Unique Levels": [
        sorted(cust_data[col].dropna().unique())
        for col in ordinal_cols
    ],
    "Number of Levels": [
        cust_data[col].nunique()
        for col in ordinal_cols
    ]
})

ordinal_summary.index = ordinal_cols

ordinal_summary

,Unique Levels,Number of Levels
Level,"[1, 2, 3]",3
HiddenScore,"[1, 2, 3, 4]",4


## Observations

The ordinal variables contain valid and well-defined ordered categories:
- Level consists of three ordered categories: 1, 2, and 3.
- HiddenScore consists of four ordered categories: 1, 2, 3, and 4.

No unexpected values or missing category levels were identified.

### Business Interpretation

Although the exact meaning of these variables is masked by the bank, they likely represent customer segmentation, risk ratings, or internal scoring systems.

Higher category values may indicate:
- Greater customer value
- Higher financial capacity
- Stronger banking relationships
- Increased loan eligibility

These hypotheses will be investigated during the Exploratory Data Analysis phase by examining their relationship with the target variable (LoanOnCard).

## Binary Feature Validation

In [20]:
# Binary Feature Summary

binary_summary = pd.DataFrame({
    "Unique Values": [
        sorted(cust_data[col].dropna().unique())
        for col in binary_cols
    ],
    "Number of Categories": [
        cust_data[col].nunique()
        for col in binary_cols
    ]
})

binary_summary.index = binary_cols

binary_summary

,Unique Values,Number of Categories
Security,"[0, 1]",2
FixedDepositAccount,"[0, 1]",2
InternetBanking,"[0, 1]",2
CreditCard,"[0, 1]",2


## Observations

The binary variables contain only the expected values:
- 0 = No
- 1 = Yes

No unexpected categories were identified.

### Business Interpretation

These features indicate whether a customer owns or uses specific banking products and services.

| Feature | Interpretation |
|----------|---------------|
| Security | Customer holds security assets with the bank |
| FixedDepositAccount | Customer has a fixed deposit account |
| InternetBanking | Customer uses internet banking services |
| CreditCard | Customer owns the bank's credit card |

#### Encoding:

| Value | Meaning |
|--------|---------|
| 0 | No |
| 1 | Yes |

These variables may provide valuable insights into customer engagement and financial behavior, which could influence loan conversion likelihood.

## Business Rule Validation

A customer's relationship duration with the bank (CustomerSince) should not exceed their age (Age). Records violating this rule are flagged as potential data quality issues and investigated before model development.

In [21]:
business_rule_checks = {
    # Customer relationship duration should not exceed the customer's age
    "CustomerSince > Age": (
        cust_data["CustomerSince"] > cust_data["Age"]
    ).sum()
}

pd.Series(business_rule_checks)

CustomerSince > Age    0
dtype: int64

## Data Understanding Summary
### Overview

The dataset was provided in two separate CSV files containing customer demographic, spending behavior, and banking product information. Both datasets were successfully merged using the unique customer identifier (ID), resulting in a consolidated dataset containing 5,000 customer records and 14 features.

The objective of this project is to develop a machine learning model that can identify customers who are likely to convert from liability customers to asset customers, enabling the bank to execute more focused and effective digital marketing campaigns.

---

### Data Structure Assessment

The dataset consists primarily of numerical variables and contains:
- 1 identifier variable (ID)
- 1 geographic variable (ZipCode)
- 5 numerical variables
- 2 ordinal variables
- 4 binary variables
- 1 target variable (LoanOnCard)

The absence of string or text-based variables simplifies preprocessing and model development.

---

### Data Quality Assessment

#### Missing Values

A missing value assessment revealed that:
- All predictor variables contain complete information.
- The target variable (LoanOnCard) contains 20 missing observations, representing approximately 0.40% of the dataset.

Since the missing values occur in the target variable, these observations will require special consideration during the data cleaning phase.

#### Duplicate Records
- No duplicate records were identified within the dataset.

Additionally, the ID variable contains 5,000 unique values, confirming that each observation represents a unique customer.

---

### Feature Assessment

#### Identifier Variable

The ID feature uniquely identifies customers and does not provide predictive information. Therefore, it will be excluded from model development.

#### Geographic Variable

The ZipCode feature contains 467 unique values, indicating high cardinality. Its predictive usefulness will be evaluated during exploratory data analysis before determining whether it should be retained, transformed, or removed.

#### Numerical Variables

The numerical features include:
- Age
- CustomerSince
- HighestSpend
- MonthlyAverageSpend
- Mortgage

These variables capture customer demographics and financial behavior and are expected to contribute significantly to predictive performance.

#### Ordinal Variables

The features Level and HiddenScore were identified as ordinal variables containing:

| Feature | Number of Levels | Description |
|----------|------------------|-------------|
| Level | 3 | Masked customer level assigned by the bank |
| HiddenScore | 4 | Masked customer score associated with the customer |

Both variables contain valid category levels and no missing values.

#### Binary Variables

The following variables were validated as binary indicators:
- Security
- FixedDepositAccount
- InternetBanking
- CreditCard

Each feature contains only valid values (0 and 1) and requires no additional validation or encoding at this stage.

---

#### Data Validation Findings

Range and validity checks were performed on all numerical variables.

Key findings include:
- No unrealistic age values were identified.
- No negative spending values were detected.
- No negative mortgage values were observed.
- A total of 52 observations contain negative values in the CustomerSince feature.

Because the variable has been masked by the bank, it is not currently possible to determine whether these values represent data quality issues or valid encoded information. Therefore, no corrective action will be taken during the Data Understanding phase.

This feature will be investigated further during exploratory data analysis.

---

### Initial Observations

Several important observations were identified during the data understanding process:
- The dataset appears generally clean and well-structured.
- Predictor variables contain complete information.
- No duplicate customer records exist.
- The target variable contains a small number of missing observations.
- The ZipCode feature exhibits high cardinality.
- Level and HiddenScore may contain valuable customer segmentation information.
- The CustomerSince feature requires additional investigation due to the presence of negative values.

---

### Conclusion

The dataset demonstrates a high level of overall quality and is suitable for machine learning model development. While a small number of missing target values and negative CustomerSince observations require further investigation, no major data integrity issues were identified.

The next phase of the project will focus on Data Cleaning and Preprocessing, where missing target observations will be addressed, feature preparation decisions will be finalized, and the dataset will be prepared for exploratory data analysis and model development.

# Phase 1 Deliverables Completed

- Business Problem Understanding
- Data Collection
- Dataset Integration
- Data Structure Inspection
- Missing Value Assessment
- Duplicate Analysis
- Unique Value Analysis
- Feature Categorization
- Range and Validity Checks
- Ordinal Variable Validation
- Binary Variable Validation
- Business Rule Validation
- Data Understanding Summary